File: backend/app/schemas/ai_analysis.py
"""Request and response schemas for the AI analysis API."""

from __future__ import annotations

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field

from app.engines.ai.models import AIAnalysisResult

AnalysisStatusLiteral = Literal["created", "running", "completed", "failed"]


class AIAnalysisRequest(BaseModel):
    """``POST /assessments/{id}/ai-analysis`` takes no parameters.

    The model exists only to *refuse* input: the evidence analysed is always
    loaded by the backend from the assessment, so a body such as
    ``{"evidence": [...]}`` is a 422, never an alternative input path.
    """

    model_config = ConfigDict(extra="forbid")


class OmittedEvidence(BaseModel):
    ref: str
    evidence_id: str | None = None
    reason: str


class AnalysisContextInfo(BaseModel):
    total_evidence: int = 0
    supplied_evidence: int = 0
    omitted: list[OmittedEvidence] = Field(default_factory=list)
    truncated_fields: int = 0
    redactions: int = 0


class AnalysisError(BaseModel):
    category: str
    message: str
    details: dict[str, Any] = Field(default_factory=dict)


class AIAnalysisResponse(BaseModel):
    """One analysis execution, as stored in ``ai_analysis_logs``."""

    id: str
    analysis_id: str
    assessment_id: str
    target_id: str | None = None
    provider: str
    model: str
    analysis_version: str
    status: AnalysisStatusLiteral
    created_at: datetime
    started_at: datetime | None = None
    completed_at: datetime | None = None
    duration_ms: int | None = None
    #: Every evidence record the model was shown (real evidence ids).
    evidence_ids: list[str] = Field(default_factory=list)
    evidence_count: int = 0
    context: AnalysisContextInfo = Field(default_factory=AnalysisContextInfo)
    usage: dict[str, int] = Field(default_factory=dict)
    result: AIAnalysisResult | None = None
    error: AnalysisError | None = None


class AIAnalysisView(BaseModel):
    """What ``GET /assessments/{id}/ai-analysis`` returns."""

    assessment_id: str
    status: Literal["not_analyzed", "created", "running", "completed", "failed"]
    #: The most recent attempt, whatever its outcome.
    latest: AIAnalysisResponse | None = None
    #: The most recent attempt that completed and passed validation.
    latest_completed: AIAnalysisResponse | None = None

File: frontend/src/types/aiAnalysis.ts
/** Mirrors backend/app/schemas/ai_analysis.py and app/engines/ai/models.py.
 *  A backend test (tests/test_frontend_contract.py) keeps the field names in step. */

export type AIAnalysisStatus = "created" | "running" | "completed" | "failed";
export type AIAnalysisViewStatus = "not_analyzed" | AIAnalysisStatus;

export type AIFindingType = "qa" | "security";
export type AIConfidence = "high" | "medium" | "low";
/** insufficient_evidence: worth noting, but the evidence does not settle it. */
export type AIFindingStatus = "supported" | "insufficient_evidence";
export type AIToolSeverity = "high" | "medium" | "low" | "informational";

export interface AIFinding {
  finding_id: string;
  type: AIFindingType;
  status: AIFindingStatus;
  title: string;
  description: string;
  impact: string;
  /** The model's confidence in its interpretation - not a severity. */
  confidence: AIConfidence;
  /** Copied from the cited evidence; validated server-side. */
  tool_severity: AIToolSeverity | null;
  /** Real evidence ids in this assessment's evidence collection. */
  evidence_ids: string[];
  /** The same records as the EV-### references the model saw. */
  evidence_refs: string[];
  /** Derived from the cited evidence, never from the model. */
  affected_components: string[];
  uncertainty: string;
}

export interface AIAnalysisResult {
  overall_assessment: string;
  findings: AIFinding[];
  evidence_gaps: string[];
  limitations: string[];
}

export interface OmittedEvidence {
  ref: string;
  evidence_id: string | null;
  reason: string;
}

export interface AnalysisContextInfo {
  total_evidence: number;
  supplied_evidence: number;
  omitted: OmittedEvidence[];
  truncated_fields: number;
  redactions: number;
}

export interface AnalysisError {
  category: string;
  message: string;
  details: Record<string, unknown>;
}

export interface AIAnalysis {
  id: string;
  analysis_id: string;
  assessment_id: string;
  target_id: string | null;
  provider: string;
  model: string;
  analysis_version: string;
  status: AIAnalysisStatus;
  created_at: string;
  started_at: string | null;
  completed_at: string | null;
  duration_ms: number | null;
  evidence_ids: string[];
  evidence_count: number;
  context: AnalysisContextInfo;
  usage: Record<string, number>;
  result: AIAnalysisResult | null;
  error: AnalysisError | null;
}

export interface AIAnalysisView {
  assessment_id: string;
  status: AIAnalysisViewStatus;
  latest: AIAnalysis | null;
  latest_completed: AIAnalysis | null;
}
